lib/bib_data_manager.rb

```diff
@@ -0,0 +1,164 @@
+require_relative 'subject_heading_data_manager'
+require_relative 'subject_component_data_manager'
+require_relative 'subfield_data_manager'
+
+class SHEP
+  class BibDataManager
+    attr_reader :institution, :bib_id, :heading_data_mgrs, :discovery_id
+
+    BIB_ID_PREFIXES_BY_INSTITUTION = Hash.new('b').merge({
+      'sierra-nypl' => 'b',
+      'recap-cul' => 'cb',
+      'recap-pul' => 'pb',
+      }).freeze
+
+      MAIN_PUBLICATION_YEAR_MARC_TAG = '008'.freeze
+      SECONDARY_PUBLICATION_YEAR_MARC_TAGS = {
+        '260' => 'c',
+        '264' => 'c',
+      }.freeze
+      # NB, SECONDARY_PUBLICATION_YEAR_MARC_TAGS.keys preserves the order above
+      PUBLICATION_YEAR_MARC_TAGS = [MAIN_PUBLICATION_YEAR_MARC_TAG] + SECONDARY_PUBLICATION_YEAR_MARC_TAGS.keys
+
+      VALID_YEAR_VALUE_RX = %r{\d[\dux-]{3}}.freeze
+
+      # according to https://neo4j.com/docs/operations-manual/current/performance/index-configuration/schema-indexes/index-key-size-calcuations/
+      MAX_STRING_LENGTH = 2018  # Half Neo4j's max index length as encoded label is also indexed & has two bytes per char
+
+      def initialize(bib_marc)
+        @bib_id = bib_marc['id']
+        @institution = bib_marc['nyplSource'] || bib_marc['nypl_source']
+
+        @discovery_id = BIB_ID_PREFIXES_BY_INSTITUTION[institution] + bib_id
+
+        if bib_marc['title'].nil?
+          error_message = "No title on bib #{bib_marc['id']}, source #{@institution}"
+          raise BibDataManagerError, error_message
+        end
+        @title = bib_marc['title'].sub(%r{\\$}, '')  # escaping terminating backslash
+
+        begin
+          @language_code = bib_marc['lang']['code']
+          bib_marc_var_fields = bib_marc['varFields'] || bib_marc['var_fields']
+          var_fields = bib_marc_var_fields.class == String ? JSON.parse(bib_marc_var_fields) : bib_marc_var_fields
+        rescue StandardError => e
+          raise BibDataManagerError, "Problem parsing JSON for #{discovery_id}: #{e.message}"
+        end
+
+        headings_fields, publication_year_fields = extract_useful_var_fields(var_fields)
+
+        @publication_year = process_publication_year publication_year_fields
+
+        @heading_data_mgrs = headings_data(headings_fields).map do |heading_data|
+          next unless subject_heading_length_ok? heading_data[:subfields]
+
+          SubjectHeadingDataManager.new_from_data(heading_data)
+        end.compact
+      end
+
+      def extract_useful_var_fields(var_fields)
+        useful = var_fields.inject({ headings: [], published_date: [] }) do |useful_fields, field|
+          next useful_fields unless field['marcTag']
+
+          useful_fields[:headings] << field if field['marcTag'].match %r{^(6\d\d)$}
+          useful_fields[:published_date] << field if PUBLICATION_YEAR_MARC_TAGS.include? field['marcTag']
+
+          useful_fields
+        end
```

Comment:
is `extract_useful_fields` the same as `[var_fields.select {|field| PUBLICATION_YEAR_MARC_TAGS.include? field['marcTag']}, var_fields.select { field['marcTag'].match %r{^(6\d\d)$} }]`